compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/DesktopTextInputService2.kt

```diff
@@ -44,19 +48,50 @@ internal class DesktopTextInputService2(
 
     private var composingText: String = ""
 
-    fun startInput(request: PlatformTextInputMethodRequest) {
+    private var ignoreNextInputMethodEvent = false
+
+    private var selectionListeningJob: Job? = null
+
+    fun startInput(request: PlatformTextInputMethodRequest, coroutineScope: CoroutineScope) {
         receivedInputMethodEventsSinceStartInput = false
+        ignoreNextInputMethodEvent = false
+
         component.enableInput(
             InputMethodSession(component, request).also {
                 inputMethodSession = it
             }
         )
+
+        // Listen to selection changes, and if a change doesn't match
+        // `inputMethodSession.expectedNextSelection`, end the composition
+        selectionListeningJob = coroutineScope.launch {
+            snapshotFlow { request.state.selection }.collect { selection ->
+                val inputMethodSession = inputMethodSession ?: return@collect
+                val expectedNextSelection = inputMethodSession.expectedNextSelection
+                inputMethodSession.expectedNextSelection = null
+
+                if (composingText.isEmpty()) return@collect
+                if ((expectedNextSelection == null) || (expectedNextSelection != selection)) {
+                    // BTF1 commits the composition itself when the user clicks, but BTF2 does not.
+                    // So this aligns the behavior of the two (a no-op for BTF1).
+                    request.editText {
+                        finishComposingText()
+                    }
+
+                    // The event sent due to ending the composition must be ignored because we've
+                    // already committed the composition.
+                    ignoreNextInputMethodEvent = true
+                    component.endComposition()
```

Comment:
```suggestion
                    SwingUtilities.invokeLater {
                        ignoreNextInputMethodEvent = true
                    }
                    component.endComposition()
```

Does `endComposition` schedule `inputMethodTextChanged` or does it call synchronously?

If the first, then we assume that the next event will be about committing. But what if there are already scheduled events, which aren't handled yet? `invokeLater` assures that there are no events between `ignoreNextInputMethodEvent = true` and `inputMethodTextChanged`.